### ISSUE TITLE
Feature: Enforce __all__ in all modules?

### ISSUE THREAD

**Issue by mgzenitech** [DetachHead/basedpyright#463]
Not sure if it's scope of basedpyright, but maybe it would be possible to implement something like __all__ validation so that there wouldn't be any modules without explicit exports? :)

**Comment by DetachHead** [comment]
interesting idea. personally i never use `__all__` and instead just rely on prefixing symbols with `_` to mark them as private

i think both ways are pretty gross, i wish there was a cleaner way to do it, something like an `@public` or `@private` decorator. but since that's not a standardized typing feature it wouldn't be an ideal solution for some people.

**Comment by mgzenitech** [comment]
> interesting idea. personally i never use `__all__` and instead just rely on prefixing symbols with `_` to mark them as private
> 
> i think both ways are pretty gross, i wish there was a cleaner way to do it, something like an `@public` or `@private` decorator. but since that's not a standardized typing feature it wouldn't be an ideal solution for some people.

agreed, but with `__all__` at least it's possible to skip `_` prefixes :) I personally only use `_` only in classes (would prefer @public, @private, @protected decorators though :D)

**Comment by DetachHead** [comment]
some thoughts i've had about this feature:

- having everything be puiblic by default and using a `@private` decorator to change that will allow this feature to work for library developers whose users aren't necessarily using basedpyright, since the decorator can have a signature that changes its type to be unusable:
  ```py
  def private(fn: T) -> Never:
      ...
  
  @private
  def foo() -> int:
      ...
  
  print(foo)
  ```
  basedpyright can then just special-case the `@private` decorator to ignore that signature when the function/class is being called from within same module
- speaking of special casing, i want to avoid it as much as possible however. instead i would rather put some thought into designing lower-level features that can be used in many different ways to allow features like this, and potentially many others, without any special casing needed. for example if we had [intersections](https://github.com/DetachHead/basedpyright/issues/170), [assertion types](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions) and some other typescript-like features, perhaps an `@public` decorator could be implemented without any special casing at all:
  ```py
  def public[T](fn: T) -> T & Asserts[type[__all__] is (*__all__, T.__name__)]:
      ...
  
  @public # calling the decorator mutates the type of __all__
  def foo():
      ...
  ```
- ideally there would be a way to mark non-functions/classes as public/private, which can't be done with decorators. (related: #48)

**Comment by mgzenitech** [comment]
I have one more idea that could be very universal... what about just adding setting to ensure that we have mandatory symbols in each module?

In my case I'd add "__all__" to enforce these. Other people may have other things that they would like to enforce, etc.

**Comment by DetachHead** [comment]
this sounds like an idea @KotlinIsland and i have had for a while, a way to enforce a module's type. maybe something like this:

```py
@module_type # tells the type checker that the exported symbols in this module must match this interface
class _Foo(Protocol):
    a: int

a = "" # error: does not match the module type
```

**Comment by KotlinIsland** [comment]
https://github.com/KotlinIsland/basedmypy/issues/262

**Comment by bluss** [comment]
@KotlinIsland I don't know exactly what you think about Python's conventions about this, but I found that pyright's author seems to disagree with how you interpret `__all__` - and with how I interpret it.

See discussion here https://github.com/microsoft/pyright/issues/3147  where it seems to say that a symbol is private if it's EITHER in `__all__` or it's not starting with an underscore.

I think https://typing.readthedocs.io/en/latest/spec/distributing.html#packaging-typed-libraries  could support my interpretation, or that's where it was informed, while pyright seems to have their own changed wording of the same rules.

If I misunderstood your standpoint on this, I'm sorry, but it might be interesting to discuss. One of the easy tests is what is the public interface of a module with `__all__ = []; def func(): pass`, is it empty or is `func` public?
